controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitTransactionTask.java

```diff
@@ -0,0 +1,174 @@
+                                 final StreamMetadataStore streamMetadataStore,
+                                 final ScheduledExecutorService executor) {
+        Preconditions.checkNotNull(streamMetadataStore);
+        Preconditions.checkNotNull(streamMetadataTasks);
+        Preconditions.checkNotNull(executor);
+        this.streamMetadataTasks = streamMetadataTasks;
+        this.streamMetadataStore = streamMetadataStore;
+        this.executor = executor;
+        this.processedEvents = null;
+    }
+
+    @Override
+    public CompletableFuture<Void> execute(CommitEvent event) {
+        // 1. get epoch.. check if the epoch matches the epoch in commit event.. if true, we can proceed.. else throw OperationNotAllowedException
+        // 2. check if txn-commit-node exists.
+        //      if node exists and has same epoch, process the node.
+        // 3. if node doesnt exist, collect all committing txn in the epoch and create txn-commit-node.
+        // 4. loop over the transactions and commit one transaction at a time until all such transactions are committed.
+        // 5. delete txn-commit-node
+        // 6. try complete scale.
+        // Note: Once scale is completed, all events for this epoch will become a no-op as all transactions would have been committed already.
+        String scope = event.getScope();
+        String stream = event.getStream();
+        int epoch = event.getEpoch();
+        OperationContext context = streamMetadataStore.createContext(scope, stream);
+        log.debug("Attempting to commit available transactions on epoch {} on stream {}/{}", event.getEpoch(), event.getScope(), event.getStream());
+
+        return streamMetadataStore.getActiveEpoch(scope, stream, context, false, executor).thenComposeAsync(pair -> {
+            if (epoch < pair.getKey()) {
+                log.debug("Epoch {} on stream {}/{} is already complete.", epoch, scope, stream);
+                return CompletableFuture.completedFuture(null);
+            } else if (epoch == pair.getKey()) {
+                // If the transaction's epoch is same as the stream's current epoch, commit it.
+                return tryCommitTransactions(scope, stream, epoch, context, this.streamMetadataTasks.retrieveDelegationToken());
+            } else {
+                // If commit request if from higher epoch, throw OperationNotAllowed. This will result in event being posted back.
+                String message = String.format("Transaction on old epoch should complete before allowing commits on new epoch %d on stream %s/%s",
+                        epoch, scope, stream);
+                throw StoreException.create(StoreException.Type.OPERATION_NOT_ALLOWED, message);
+            }
+        }).whenCompleteAsync((result, error) -> {
+            if (error != null) {
+                log.error("Exception while attempting to committ transaction on epoch {} on stream {}/{}", epoch, scope, stream);
```

Comment:
log the error too

Reply:
Also how is this handled? If it gets here, then nobody else will know an error occurred.

Reply:
this is because `whenComplete` callback does not change the returned future's result and exception. `handle` does.